src/type-extension.ts

```diff
@@ -45,9 +45,19 @@ export default abstract class LuaTypeExtension<T, K extends BaseDecorationOption
             throw new Error(`metatable not found: ${this.name}`)
         }
 
-        // Set as the metatable for the userdata.
-        // -1 is the metatable, -2 is the user data.
-        thread.cmodule.lua_setmetatable(thread.address, -2)
+        if (decorations?.metatable) {
+            // If we already have a gc handler, let's keep, otherwise inherit it
+            if (!decorations.metatable.__gc) {
+                const gcType = thread.cmodule.lua_getfield(thread.address, -1, '__gc')
+                if (LuaType.Function === gcType) {
+                    decorations.metatable.__gc = decorate<any>(thread.getPointer(-1), { functionPointer: true })
```

Comment:
If you replace the metatable I think it'd be a good idea to make the new one read-only to avoid the __gc field being deleted

Reply:
I think this is a difficult problem (replacing the metatables and not losing memory). I don't think there's really a safe way to do it.
I think you could very easily automate what I demo'd in #26. Perhaps by supply a wrapClass function? You could automate it further by adding a type extension that when  { wrap: true } is passed it auto-applies that wrapClass?

Reply:
> If you replace the metatable I think it'd be a good idea to make the new one read-only to avoid the __gc field being deleted

If a metatable is being passed, there is no way to know if the user want to get it later, making it read-only can be a problem.

Reply:
> I think this is a difficult problem (replacing the metatables and not losing memory). I don't think there's really a safe way to do it.

I was thinking in make the `LuaTypeExtension` all abstract and pass the code to `UserdataTypeExtension`, with it we can handle everything in there, it will be easy to solve the memory issue.

Reply:
Did my other suggestion not work for your use case?

The abstract thing could work but there is two different types of things that need cleaning up and having them both managed in the abstract class feels like a step backwards to how Thread handled it before type extensions.

Another option you may not have considered is dynamically adding and removing type extensions?

Reply:
> Did my other suggestion not work for your use case?

Probably, but I can't think in a way of do that in the library without increasing a lot the complexity and glue code. Maybe I'm not seeing what you're expressing.

> Another option you may not have considered is dynamically adding and removing type extensions?

I don't know how that would work 

Another option that I am considering for the PR code not to waste memory is to share a default gcPointer between the types, and if you need a different to your type, just override it.

Reply:
I've made a PR onto demoon that converts it to use the same structure as my demo in test.engine.js.
https://github.com/ceifa/demoon/pull/1

Tbh, I'm not super sure either, if you still don't like option 1 (shown in the PR) then I'll have a go at this.

I'm not sure that would work in the base class because it uses the metatable name to check if it's the right sort to free and during isType to figure out if it's the right type. You could setup a default jsref GC in the base classes constructor though.

Reply:
> I've made a PR onto demoon that converts it to use the same structure as my demo in test.engine.js.
> [ceifa/demoon#1](https://github.com/ceifa/demoon/pull/1)
> Tbh, I'm not super sure either, if you still don't like option 1 (shown in the PR) then I'll have a go at this.

As a library user, idk if I was very comfortable with this workaround, would be nice to have it working out of the box.

> I'm not sure that would work in the base class because it uses the metatable name to check if it's the right sort to free and during isType to figure out if it's the right type. You could setup a default jsref GC in the base classes constructor though.

Makes sense, I will take a look and make a test for this

Reply:
Is the proxy in this PR suitable for demoon? https://github.com/ceifa/wasmoon/pull/28

Reply:
Yes! It looks very nice :)